src/input/plugins/YoutubeInputPlugin.cxx

```diff
@@ -0,0 +1,182 @@
+/*
+ * Copyright 2003-2020 The Music Player Daemon Project
+ * http://www.musicpd.org
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License along
+ * with this program; if not, write to the Free Software Foundation, Inc.,
+ * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+ */
+
+#include "YoutubeInputPlugin.hxx"
+#include "CurlInputPlugin.hxx"
+#include "PluginUnavailable.hxx"
+#include "../InputPlugin.hxx"
+#include "../InputStream.hxx"
+#include "../TaggedInputStream.hxx"
+#include "Chrono.hxx"
+#include "Log.hxx"
+#include "tag/Builder.hxx"
+#include "util/ScopeExit.hxx"
+#include "util/ExecOpen.hxx"
+#include "util/Domain.hxx"
+
+#include <yajl/yajl_tree.h>
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+
+static Domain youtube_domain("youtube");
+
+static const char *input_youtube_prefixes[] = {
+	"https",
+	nullptr
+};
+
+static void
+input_youtube_init(EventLoop &, const ConfigBlock &)
+{
+	if(WEXITSTATUS(system("youtube-dl --version > /dev/null")) != 0)
+		throw PluginUnavailable("youtube-dl not found");
+}
+
+static InputStreamPtr
+input_youtube_open(const char *uri, Mutex &mutex)
+{
+	const char *args[] = {
+		"youtube-dl",
+		"--no-playlist",
+		"--format=bestaudio",
+		"--dump-single-json",
+		uri,
+		nullptr
+	};
+
+	int pid;
+	FILE *stream = exec_open(&pid, "youtube-dl", args);
+	if(!stream) {
+		LogErrno(youtube_domain, "Can't spawn youtube-dl");
+		return nullptr;
+	}
+
+	/* Read youtube-dl output */
+	std::string json;
+	char buf[512];
+	ssize_t size = 0;
+	while(true) {
+		size = fread(buf, sizeof(char), 512, stream);
+		if(size > 0)
+			json.append(buf, size);
```

Comment:
What if the process just happens to print a few terabytes of data? This will crash MPD.